### PR TITLE
async: avoid full sweeps if possible

### PR DESCRIPTION
**Pull request by indutny**
/cc @bnoordhuis @piscisaureus

### CODE REVIEW

File: src/unix/async.c [hunk 1]
```diff
@@ -81,11 +81,16 @@ int uv_async_send(uv_async_t* handle) {
     return 0; /* already pending */
 
   do
-    r = write(handle->loop->async_pipefd[1], "x", 1);
+    r = write(handle->loop->async_pipefd[1], (char*) &handle, sizeof(&handle));
```
Comment (bnoordhuis) [inline src/unix/async.c]: You're taking the address of a local variable...

Comment (indutny) [inline src/unix/async.c]: No, I'm writing data from this local address.

File: src/unix/async.c [hunk 2]
```diff
@@ -81,11 +81,16 @@ int uv_async_send(uv_async_t* handle) {
...
   while (r == -1 && errno == EINTR);
 
-  if (r == -1 && errno != EAGAIN && errno != EWOULDBLOCK)
-    return uv__set_sys_error(handle->loop, errno);
+  if (r == -1) {
+    /* If pipe is full, event loop should sweep all handles */
+    if (errno == EAGAIN || errno == EWOULDBLOCK)
+      handle->loop->async_sweep_needed = 1;
+    else
+      return uv__set_sys_error(handle->loop, errno);
```
Comment (bnoordhuis) [inline src/unix/async.c]: Why not abort()? Anything else is a fatal error anyway, right?

Comment (indutny) [inline src/unix/async.c]: What else? :) There're no abort() calls in this function. Also, that's how it was before my patch.

Comment (bnoordhuis) [inline src/unix/async.c]: Fair enough. We can change that some other time.

**Comment by indutny** [comment]
Benchmarks are now working fine (w/o any assertions), additionaly this patch doesn't allow callbacks to be invoked by closed async handles.

File: src/unix/async.c (hunk 2 above)
Comment (piscisaureus) [inline src/unix/async.c]: Race condition: what if the following happens:
- Worker thread: write() EAGAIN
- Main thread: drain pipe
- Worker thread: set async_sweep_needed

I suppose you should retry the write after you set the async_sweep_needed bit.

Comment (indutny) [inline src/unix/async.c]: You're right, fixed.

**Pull request closed without merging**